src/pythia/src/utils/web3.rs

```diff
@@ -0,0 +1,197 @@
+use std::str::FromStr;
+
+use anyhow::{Context, Result};
+
+use candid::Nat;
+
+use ic_cdk::api::management_canister::http_request::{TransformContext, TransformFunc};
+use ic_web3_rs::{
+    ic::KeyInfo,
+    transports::{ic_http_client::CallOptionsBuilder, ICHttp},
+    types::{Transaction, TransactionId, TransactionParameters, TransactionReceipt, H256},
+    Transport, Web3,
+};
+
+use super::{address, canister, nat, time, web3};
+use crate::{
+    clone_with_state, retry_until_success,
+    types::{chains::Chains, errors::PythiaError},
+};
+
+const ECDSA_SIGN_CYCLES: u64 = 23_000_000_000;
+const TRANSFER_GAS_LIMIT: u64 = 21_000;
+const TX_SUCCESS_STATUS: u64 = 1;
+
+pub fn instance(chain_id: &Nat) -> Result<Web3<ICHttp>> {
+    Ok(Web3::new(ICHttp::new(&Chains::get(chain_id)?.rpc, None)?))
+}
+
+pub async fn get_tx(chain_id: &Nat, tx_hash: &str) -> Result<Transaction> {
+    let tx_hash = H256::from_str(tx_hash)?;
+    let w3 = instance(chain_id)?;
+
+    let tx_receipt = retry_until_success!(w3
+        .eth()
+        .transaction_receipt(tx_hash, canister::transform_ctx_tx_with_logs()))?
+    .context(PythiaError::TxDoesNotExist)?;
+
+    match tx_receipt.status {
+        Some(status) => {
+            if status.as_u64() != 1 {
+                return Err(PythiaError::TxHasFailed.into());
+            }
+        }
+        None => return Err(PythiaError::TxNotExecuted.into()),
+    }
+
+    retry_until_success!(w3
+        .eth()
+        .transaction(TransactionId::from(tx_hash), canister::transform_ctx_tx()))?
+    .context(PythiaError::TxDoesNotExist)
+}
+
+pub async fn gas_price(chain_id: &Nat) -> Result<Nat> {
+    let w3 = instance(chain_id)?;
+    Ok(nat::from_u256(&retry_until_success!(w3
+        .eth()
+        .gas_price(canister::transform_ctx()))?))
+}
+
+pub fn key_info() -> KeyInfo {
```

Comment:
You are using this as a const value of `KeyInfo`, as I see, so at least make in #[inline], as max add a `const KEY_INFO: KeyInfo` value